sky/serve/serve_utils.py

```diff
@@ -719,6 +731,160 @@ def _get_replica_status() -> serve_state.ReplicaStatus:
     return ''
 
 
+def has_valid_replica_id(file_name: str,
+                         target_replica_id: Optional[int]) -> bool:
+    if target_replica_id is None:
+        return True
+
+    match = re.search(serve_constants.REPLICA_ID_PATTERN, file_name)
+
+    if match is not None:
+        replica_id = int(match.group(1))
+    else:
+        raise ValueError(
+            f'Failed to get replica id from file name: {file_name}')
+
+    return replica_id == target_replica_id
+
+
+def append_job_logs_to_replica_log(new_replica_log_file: str,
+                                   job_log_file_name: Optional[str],
+                                   replica_id: int) -> None:
+
+    if job_log_file_name is not None:
+        logger.info(f'\n== End of logs (Replica: {replica_id}) ==')
+        with open(new_replica_log_file, 'a',
+                  encoding='utf-8') as replica_log_file, open(
+                      job_log_file_name, 'r', encoding='utf-8') as job_file:
+            replica_log_file.write(job_file.read())
+        os.remove(job_log_file_name)
+    else:
+        with open(new_replica_log_file, 'a',
+                  encoding='utf-8') as replica_log_file:
+            replica_log_file.write('Failed to sync down job logs from replica '
+                                   f'{replica_id}.\n')
+
+
+def copy_log_file(log_file_name: str, replica_id: Optional[int],
+                  dir_for_download: str,
+                  target_replica_id: Optional[int]) -> bool:
+    if has_valid_replica_id(
+            log_file_name, target_replica_id) and os.path.exists(log_file_name):
+        shutil.copy(log_file_name, dir_for_download)
+        return True
+    else:
+        logger.error(
+            f'{colorama.Fore.RED}Path {dir_for_download} does not exist, '
+            f'or replica_id {replica_id} is invalid.'
+            f'{colorama.Style.RESET_ALL}')
+        return False
+
+
+def sync_normal_replica_logs(service_name: str, replica: dict,
+                             dir_for_download: str,
+                             target_replica_id: Optional[int]) -> None:
```

Comment:
Not sure what is a "normal" replica. Do you mean nonterminal replica?
Also, this function is only called once. Lets inline it to reduce fragmentation.

Reply:
Yes, it means nonterminal replica here

In fact, I'm a little bit confused about the relationship between inlining and redundancy here. If all these helper functions above are put back in their original place, it will make the main function `sync_down_logs()` / `prepare_replica_logs_for_download()` quite long and the logic rather complex. I'm not sure if I should inline them?